### DISCUSSION TITLE
ARM .bin code Emulation

### DISCUSSION THREAD

**Discussion by a-michelis** [unicorn-engine/unicorn#1887] [Q&A]
Hello, everyone!

I'm trying to create a tool that emulates ARM code compiled for a specific SOC -I'm at the stage of research, at the moment.

From the information i gathered so far, in order to emulate ARM code designed for an SOC, I have to:
- Configure properly _uc_ to run with ARM code
- Implement All (needed) SOC peripherals and properly map them to their respective regions
- Implement and map all registers
- Map the software's sections to their respective memory regions

The last step is the one I'm having the most trouble with:
Creating an `.elf` file loader doesn't seem to be an extremely difficult task, while the information an elf  it provides allows me to discover all regions and map them properly. 

What if i want to run a `.bin` file, tho?
What are the steps I should take to reach my end goal of emulating the software?

Thank you in advance!

*PS: If I missed steps, feel free to tell me, so I can further advance my understanding!

**Comment by wtdcode** [comment]
I'm not closely looking at the discussion because they don't appear in my notification. Sorry for they delay.

You should have a look at how Qiling does this: https://github.com/qilingframework/qiling/
